Metaknob soft takeover

Be-ing opened the pull request:
Follow-up from discussion on #1062:
* When an effect is on, do not make parameters audibly jump when metaknob changes. If the superknob is turned, do not make metaknobs jump for effects that are on.
* When an effect is off, do not do any automatic soft takeover. Snap parameters to position of metaknob when the metaknob turns (either directly or by the superknob). With #1103, this does not interfere with loading initial states of knobs from controllers when Mixxx starts. It also makes it easier to experiment with different metaknob linkings.

Be-ing committed:
implement super -> metaknob and metaknob -> parameter soft takeover

If an effect is on, avoid making parameters audibly jump. If effect is
off, do not do soft takeover, just snap parameters to position of
metaknob. Controller mappings still must enable soft takeover for
the super/metaknobs as appropriate.

daschuer commented:
This works as advertised, and seams to be nice. I there are only some remaining worries, if the conditional disabled oversoft will leads to confusion because it may kick in or out when not expected. 
A third voice would be nice. 
@timrae: Could you have a look?

daschuer commented:
I wonder why we see commits here that are currently in master. 
Did you debase the previous branch? Than this need to be rebase on master as well.

timrae commented:
> @timrae: Could you have a look?

Sure

Be-ing commented:
Yikes I tried to rebase and it lost the new commit >.<

Be-ing commented:
Alright, looks good now. I think that happened because I branched off my per_effect_metaknobs branch, added a commit to this branch, then added another commit to per_effect_metaknobs before that was merged.

timrae commented on src/effects/effectslot.cpp:
```diff
@@ -63,6 +63,8 @@ EffectSlot::EffectSlot(const QString& group,
     connect(m_pControlMetaParameter, SIGNAL(valueChanged(double)),
             this, SLOT(slotEffectMetaParameter(double)));
 
+    m_pSoftTakeover = new SoftTakeover();
```
Does this need a `delete` or it owned by the object tree?

timrae commented:
I'm testing it now... With the current master branch, metaknob itself has soft takeover already (I only checked flanger which only has one parameters assigned to meta, it may be different for other effects). However, the super knob and the meta knob do not have any soft takeover in terms of the interaction between them. Is that the objective of this PR?

The sync between the two is still not perfect, as demonstrated by the following steps:

1) Enable flanger effect
2) Turn metaknob all the way down to zero
3) Turn super knob all the way up to max
4) Turn metaknob

--> Audible jump

timrae commented:
> the conditional disabled oversoft will leads to confusion because it may kick in or out when not expected

Do you mean people will be confused that the soft takeover doesn't occur when the effect is disabled? The only situation where I think it might be confusing is if the user maps both super knob and meta knobs on their controller. We should tell them not to do this.

Otherwise, in general I feel that soft takeover being disabled is more intuitive than being enabled, so only enabling soft takeover when the effect itself is enabled (with the priority of preventing audible jumps) makes sense to me.

daschuer commented:
@timrae Thank, you. This means we are on the right track. :-)

Be-ing commented:
> The sync between the two is still not perfect, as demonstrated by the following steps:
    Enable flanger effect
    Turn metaknob all the way down to zero on your controller
    Turn super knob all the way up to max in the GUI
    Turn metaknob on your controller
--> Audible jump

I cannot reproduce that with this branch. Are you sure you tested the correct branch?

timrae commented:
I forked master branch and cherry picked this commit.

On 10 Jan. 2017 03:42, "Be" <notifications@github.com> wrote:

> The sync between the two is still not perfect, as demonstrated by the
> following steps:
> Enable flanger effect
> Turn metaknob all the way down to zero on your controller
> Turn super knob all the way up to max in the GUI
> Turn metaknob on your controller
> --> Audible jump
>
> I cannot reproduce that with this branch. Are you sure you tested the
> correct branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1111#issuecomment-271367830>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACsA4nmQL6H4a11c4Ocv6iY5CvVDRHH8ks5rQn-DgaJpZM4Ldffr>
> .
>

Be-ing commented:
Could you try checking out this branch and building it?

timrae commented:
The way I was checking it is more indicative of the effect of merging this PR, so I prefer you rebase this into master and check that. Plus I might not have time today

Be-ing commented:
Okay, rebased it onto master.

timrae commented:
Did you try to reproduce on the rebased branch?

On 11 Jan. 2017 08:57, "Be" <notifications@github.com> wrote:

> Okay, rebased it onto master.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1111#issuecomment-271736525>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACsA4vL4NknFWdYsATVs4SeVLQOv9Azhks5rRBrhgaJpZM4Ldffr>
> .
>

Be-ing commented:
Yes. I cannot reproduce the issue. It's working as intended as far as I can tell.

timrae commented:
Ok I'll take a video when I get home after work

On 11 Jan. 2017 9:49 am, "Be" <notifications@github.com> wrote:

> Yes. I cannot reproduce the issue. It's working as intended as far as I
> can tell.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1111#issuecomment-271745135>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACsA4jtaPLXZ3moLjjw_g9sXibRqKzoEks5rRCcngaJpZM4Ldffr>
> .
>

timrae commented:
It's still happening...

timrae commented:
See around 25s
[recording.zip](https://github.com/mixxxdj/mixxx/files/698904/recording.zip)

timrae commented:
This is with 42d3262 and gcc 6.2 on Ubuntu x64, Qt version 4.8.7

Be-ing commented:
Oh, I was testing with my controller which had soft takeover enabled by its script. >.<

daschuer commented:
I have watched the recording, and I actually think this behavior is correct if you do not have softtakover enabled on your controller. you most likely see an out of sync situation between GUI meta knob and controller meta knob. 
* You turn the meta knob fully left.
* Than you turn the super knob fully right. 
-> the hidden GUI Metaknob follows the super knob, the controller knob is unchanged 
* now you touch the Metaknob on the controller, the GUI knob jumps back. 
This is IMHO correct for no controller softtakover. 

The issue targeted here is just the other way round.
* Turn the super knob
-> meta follows 
* Turn the meta knob
* Turn the super knob again 
Without this PR meta jumps to super without the chance to fix this in the controller mapping
Now we get a out of sync and snap in behavior. 

Does this work for you? 
An alternative would be to let meta not follow super. But then we loose the nice GUI feedback.

Be-ing commented:
> The issue targeted here is just the other way round.

Yes, this PR fixes the issue you outlined. Together with soft takeover activated by the controller mapping it solves the issue @timrae reported.

> An alternative would be to let meta not follow super.

That would be overcomplicated. I don't think we need to introduce any more ControlObjects for the new effects interface.

Be-ing commented on src/effects/effectslot.h:
```diff
@@ -108,6 +109,8 @@ class EffectSlot : public QObject {
     QList<EffectParameterSlotPointer> m_parameters;
     QList<EffectButtonParameterSlotPointer> m_buttonParameters;
 
+    SoftTakeover* m_pSoftTakeover;
```
Does this need to be deleted in the destructor?

daschuer replied:
yes

rryan replied:
Yes, though you can make it a member variable so you don't have to allocate it at all.

Be-ing replied:
Fixed in rebased and force pushed commit.

timrae commented:
Sounds good to me

On 12 Jan. 2017 9:07 am, "Be" <notifications@github.com> wrote:

> *@Be-ing* commented on this pull request.
> ------------------------------
>
> In src/effects/effectslot.h
> <https://github.com/mixxxdj/mixxx/pull/1111#pullrequestreview-16265838>:
>
> > @@ -108,6 +109,8 @@ class EffectSlot : public QObject {
>      QList<EffectParameterSlotPointer> m_parameters;
>      QList<EffectButtonParameterSlotPointer> m_buttonParameters;
>
> +    SoftTakeover* m_pSoftTakeover;
>
> Does this need to be deleted in the destructor?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1111#pullrequestreview-16265838>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACsA4gSJF4LspUMv_-hE5XtPtjvGcH8aks5rRW6kgaJpZM4Ldffr>
> .
>

Be-ing commented:
The other issue this PR solves is:
1. Turn the metaknob
1. Adjust a parameter knob linked to the metaknob
2. Turn the metaknob
If the effect is enabled, the manually adjusted parameter should not jump.

Be-ing commented:
Ready for merge?

rryan commented:
Looks good to me. SuperLinkTest.Softtakeover is failing because the test does not enable the effect. Could you split SuperLinkTest.SoftTakeover into SuperLinkTest.SoftTakeover_EffectEnabled and SuperLinkTest.SoftTakeover_EffectDisabled to cover both cases?

```
[0;32m[ RUN      ] [mSuperLinkTest.Softtakeover
Loading resources from  "/home/travis/build/mixxxdj/mixxx/res/" 
src/test/super_link_test.cpp:120: Failure
Value of: m_pControlValue->get()
  Actual: 0.25
Expected: 1.0
Which is: 1
src/test/super_link_test.cpp:128: Failure
Value of: m_pControlValue->get()
  Actual: 0.55
Expected: 0.1
[0;31m[  FAILED  ] [mSuperLinkTest.Softtakeover (9 ms)
```

Be-ing committed:
update metaknob link soft takeover tests

Test for new metaknob -> parameter and superknob -> metaknob behavior both when effect is enabled and disabled.

rryan commented:
Nice, thanks for the extra tests!

Be-ing commented:
Is there a such thing as extra tests? :P
